WT-11178 Detect bitmap value change for incremental backups

jeremythorp opened the pull request:
When tracking blocks for incremental backup, the set of changed blocks is tracked via a bitmap in the code. There is an invariant that single bits should never be cleared from the block tracking bitmap - they should only be added. This ticket adds checks that work across a WiredTiger restart to detect any bits incorrectly getting changed from 1 to 0.

ddanderson commented on src/meta/meta_ckpt.c:
```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
     return (ret);
 }
 
+/*
+ * ckpt_blkmod_to_item --
+ *     Extracts the blkmod info in a checkpoint into a bitmap in a WT_ITEM.
+ */
+static int
+ckpt_blkmod_to_item(WT_SESSION_IMPL *session, WT_CKPT *ckpt, const char *id, WT_ITEM *output_item)
```
I wonder if these functions, plus the new changes to `__wt_meta_ckptlist_set` might better belong in `cur_backup_incr.c`?  Just because they have similarities.  Or maybe a new file `meta_backup_incr.c`?   I'm not committed to any one of these, just throwing out for consideration.

jeremythorp replied:
Hi @ddanderson , that's definitely worth considering. I put `ckpt_blkmod_to_item` in `meta_ckpt.c` because the only place it is currently used is in `__wt_meta_ckptlist_set`, but if there's value in moving it, we can do that.

ddanderson commented on src/meta/meta_ckpt.c:
```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
+    metadata_cursor = NULL;
+    file_config = NULL;
+
+    WT_RET(__wt_metadata_cursor_open(session, NULL, &metadata_cursor));
```
This function opens the metadata and positions at the uri.  But isn't it called from somewhere up the stack where we are positioned at the same entry in the metadata?  (I could be wrong). If that's the case, maybe the needed metadata can be passed down into this function rather than doing another lookup.

jeremythorp replied:
I've had a look at the call stack and I don't think the metadata is always available to the calling code, but I may have missed something.

ddanderson replied:
Okay, thanks for checking.

ddanderson reviewed (commented):
Basically this looks good, I noted several things to consider to improve.

marcbutler commented on test/unittest/tests/helpers/vector_bool.cpp:
```diff
@@ -0,0 +1,139 @@
+}
+
+std::vector<bool>
+operator&(std::vector<bool> const &a, std::vector<bool> const &b)
```
This operation semantic is similar but different to that of integral values, for example:
```
0x1 & 0xFF == 0x1 (integer)
0x1 & 0xFF == 0xFF (vector<bool>)
```
Which may be surprising. I think this needs to be considered. A possibility is to make the operator invalid for  `a.size() != b.size()`.

jeremythorp replied:
I've added a test case that shows that 0x01 & 0xff == 0x01 as expected. 

```
        std::vector<bool> v1(vector_bool_from_hex_string("1"));
        std::vector<bool> v2(vector_bool_from_hex_string("ff"));

        REQUIRE(get_true_count(v1) == 1);
        REQUIRE(get_true_count(v2) == 8);

        std::vector<bool> v_and(v1 & v2);
        std::vector<bool> v_xor(v1 ^ v2);

        REQUIRE(v_and == vector_bool_from_hex_string("01"));
        REQUIRE(v_xor == vector_bool_from_hex_string("fe"));
```

marcbutler replied:
@jeremythorp Thanks for that, I'm still curious about the following behavior:
```
    Section("Behavior Question")
    {
        std::vector<bool> v1(vector_bool_from_hex_string("1"));
        std::vector<bool> v2;
        std::vector<bool> v3;

        REQUIRE(get_true_count(v1) == 1);
        REQUIRE(get_true_count(v2) == 0);
        REQUIRE(get_true_count(v3) == 0);

        std::vector<bool> v1_and_v2(v1 & v2);
        std::vector<bool> v2_and_v3(v2 & v3);

        REQUIRE(get_true_count(v2_and_v3) == 0);
        REQUIRE(v2_and_v3.size() == 0);
        REQUIRE(get_true_count(v1_and_v2) == 0);
        REQUIRE(v1_and_v2.size() == 0);
    }
```
The last test `REQUIRE(v1_and_v2.size() == 0);` fails. It's not a logical bug, but the behavior is different.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
+    WT_RET(__wt_metadata_cursor_open(session, NULL, &metadata_cursor));
+
+    metadata_cursor->set_key(metadata_cursor, uri);
+    WT_RET(metadata_cursor->search(metadata_cursor));
+    WT_RET(metadata_cursor->get_value(metadata_cursor, &file_config));
+    WT_RET(__wt_config_getones(session, file_config, "checkpoint_backup_info", &value));
```
That's a good idea.

jeremythorp replied:
Done!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -885,6 +885,7 @@ __wt_meta_ckptlist_get_from_config(WT_SESSION_IMPL *session, bool update, WT_CKP
              * Load any incremental information from the configuration string into the existing base
              * checkpoints.
              */
+            /* Comment out the next line to recreate the incremental bitmap backup bug */
```
Done!

jeremythorp replied:
It was useful having it there when debugging the code, but it's better to get rid of it now.

jeremythorp commented on test/csuite/incr_backup/main.c:
```diff
@@ -807,69 +807,39 @@ check_backup(const char *backup_home, const char *backup_check, TABLE_INFO *tinf
+    /* Save the rnd state in the seed global variable for error reporting */
+    seed = rnd->v;
 
     preserve = false;
```
Well spotted - `preserve` is used, but it shouldn't be overwriting the parameter.

jeremythorp replied:
Done!

sueloverso commented:
@jeremythorp there are several comments where you replied "done" but the code does not look changed. Can you push the latest changes and I will take a look?

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
+
+    for (index = 0; index < original_bitmap->size && (*ok); index++) {
+        /* Detect bits that were (incorrectly) changed from 1 in the original to 0 in the new */
+        partial_result = (*original_ptr ^ *new_ptr) & *original_ptr;
```
You're right - I've simplified the expression.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
+        partial_result = (*original_ptr ^ *new_ptr) & *original_ptr;
+
+        if (partial_result != 0)
+            *ok = false;
```
Thanks for the suggestion - I've switched to a break.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
+                        WT_RET(__wt_nhex_to_raw(session, blocks.str, blocks.len, output_item));
+                }
+            }
+        }
```
The code in cur_backup_incr.c/__curbackup_incr_blkmod() also has a loop through the config values which starts with the following line:

`while ((ret = __wt_config_next(&blkconf, &k, &v)) == 0) {`

In that case the loop is longer than in the loop in get_blkmods().

So, I think the pattern in get_blkmods() is similar (but shorter and simpler) than that in __curbackup_incr_blkmod().

I agree the new code could be made clearer, so I've refactored it and added comments.

jeremythorp commented on test/unittest/tests/helpers/vector_bool.cpp:
```diff
@@ -0,0 +1,139 @@
+    std::string result;
+
+    auto iter = vector_bool.cbegin();
+    while (iter != vector_bool.cend()) {
```
Thanks for the suggestion. I've removed the inner loop and it simplifies the code.

jeremythorp commented on test/unittest/tests/helpers/vector_bool.cpp:
```diff
@@ -0,0 +1,139 @@
+
+    for (auto iter = hex_str.crbegin(); iter != hex_str.crend(); ++iter) {
+        int value_of_hex_char = std::stoi(std::string(1, *iter), nullptr, 16);
+        result.push_back((value_of_hex_char & 1) != 0);
```
The loop performs a reverse iteration through the string so that bits are appended (rather than prepended) to the bitmap. I've added a comment to that effect, and hopefully it's now clearer.

I've also added some extra test cases to confirm that 0x3c is encoded as 0011 1100 (note the bitmap initialisers are least significant bit first).

```
    test_hex_conversion({false, false, false, false, true, false, false, false}, "10");
    test_hex_conversion({true, false, false, false, false, false, false, false}, "01");
    test_hex_conversion({false, false, true, true, true, true, false, false}, "3c");
    test_hex_conversion({true, true, false, false, false, false, true, true}, "c3");
```

marcbutler replied:
Hi Jeremy, I'm sorry but I still don't understand why you're reversing the bits, according to your tests quoted above:
```
0x10 == 0b00001000
0x01 == 0b10000000
```
But my expectation is:
```
0x10 == 0b00001000
0x01 == 0b00000001
```
Here is a demonstration in c++17:
```
ip-10-0-101-50:tmp [1]$ cat binary.cc
// clang -std=c++17 -o binary binary.cc
#include <iostream>
#include <iomanip>
int main(void) {
	using namespace std;
	unsigned char b1 = 0b00010000;
	unsigned char b2 = 0b00000001;
	cout << setw(2) << hex <<
		"b1 = " << unsigned(b1) << endl <<
		"b2 = " << unsigned(b2) << endl;
	return 0;
}

$ ./binary
b1 = 10
b2 = 1
```

jeremythorp commented:
Hi @sueloverso , thanks very much. I've pushed my changes, and I'd welcome your feedback.

sueloverso commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
+
+                        if ((ret != 0) || !blkmods_are_ok)
+                            WT_ERR_MSG(session, EINVAL,
+                              "File blkmods are not compatible with those in the checkpoint");
```
I think the other function should assert and panic and drop core at the moment the problem is detected. I don't think it should be an error return.

Also should this be under HAVE_DIAGNOSTIC? I don't remember what was discussed originally but I'd be concerned that for a lot of tables on a release build it may add checkpoint processing time.

sueloverso replied:
I am fine with it not being DIAGNOSTIC unless we see perf impact.

sueloverso commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
     WT_DECL_RET;
-    const char *fname;
-    bool has_lsn;
+    WT_ITEM file_blkmods_buffer, checkpoint_blkmods_buffer;
```
Alphabetize.

jeremythorp replied:
Done!

sueloverso commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
-        if (F_ISSET(ckpt, WT_CKPT_ADD))
+        if (F_ISSET(ckpt, WT_CKPT_ADD)) {
+            WT_CLEAR(file_blkmods_buffer);
+            WT_CLEAR(checkpoint_blkmods_buffer);
```
Alphabetize.

jeremythorp replied:
Done!

sueloverso reviewed (commented):
I have a lot of comments and suggestions. I want to take another look at this.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,151 @@ __wt_meta_ckptlist_update_config(
 
+/*
+ * ckpt_blkmod_to_item --
+ *     Extracts the blkmod info in a checkpoint into a bitmap in a WT_ITEM.
```
That makes sense - I'll make the change.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,152 @@ __wt_meta_ckptlist_update_config(
+     */
+    for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk) {
+        if (!F_ISSET(blk, WT_BLOCK_MODS_VALID))
+            continue;
```
That's a good point! Given the structure of the second loop, the first loop and associated code is redundant can be removed.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
     WT_CKPT_FOREACH (ckptbase, ckpt)
-        if (F_ISSET(ckpt, WT_CKPT_ADD))
+        if (F_ISSET(ckpt, WT_CKPT_ADD)) {
+            WT_CLEAR(file_blkmods_buffer);
+            WT_CLEAR(checkpoint_blkmods_buffer);
```
Good point - I'll make the change.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
+                    WT_ERR(
+                      ckpt_blkmod_to_item(session, ckpt, blk->id_str, &checkpoint_blkmods_buffer));
+
+                    ret = get_blkmods(session, filename, blk->id_str, &file_blkmods_buffer);
```
ret should be checked - I've replaced the assignment to ret with a WT_ERR()

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,152 @@ __wt_meta_ckptlist_update_config(
+
+    for (index = 0; index < original_bitmap->size; index++) {
+        /* Detect bits that were (incorrectly) changed from 1 in the original to 0 in the new */
+        partial_result = *original_ptr & ~(*new_ptr);
```
I'm absolutely all for using tried and true well-tested functions whenever possible. However in this case, I think that the proposed code

```
    for (index = 0; index < original_bitmap->size; index++) {
        /* Detect bits that were (incorrectly) changed from 1 in the original to 0 in the new */
        partial_result = *original_ptr & ~(*new_ptr);

        if (partial_result != 0) {
            *ok = false;
            break;
        }

        original_ptr++;
        new_ptr++;
    }
```

is simpler (and possibly faster) than using __bit_ffc(). What do you think?

sueloverso replied:
Yes, I agree that your code is much shorter and simpler. Can you elaborate on the comment about detecting the bits using bitwise complement (so that readers less familiar with C's bitwise operations have a phrase to search).

jeremythorp replied:
Hi @sueloverso , I've added a comment to describe this in more detail. How does that look now?

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
+
             WT_ERR(__wt_ckpt_blkmod_to_meta(session, buf, ckpt));
 
+            if (strcmp(filename, WT_HS_URI) == 0 || session->hs_cursor_counter != 0) {
```
We need to avoid hitting the following assert in __wt_open_cursor():

ddanderson replied:
Are we losing information by skipping files in this situation?  That is, `_wt_meta_ckptlist_set` on file X can apparently be called while a history store cursor is open.  If we skip either the collection of bits or checking of bits when that happens, is that a bug or just unfortunate?

Since we are opening a metadata cursor (readonly presumably) that should be safe, as the metadata doesn't participate in history store.  So I think we could change `__wt_open_cursor` as:
```
     * There are three exceptions to this rule:
     *  - Verifying the metadata through an internal session.
     *  - The btree is being verified.
     *  - Opening the metadata file itself.       /* ADDED */
     */
    WT_ASSERT(session,
      strcmp(uri, WT_HS_URI) == 0 || 
      strcmp(uri, WT_METADATA_URI) == 0 ||    /* ADDED */
      session->hs_cursor_counter == 0 ||
        F_ISSET(session, WT_SESSION_INTERNAL) ||
        (S2BT_SAFE(session) != NULL && F_ISSET(S2BT(session), WT_BTREE_VERIFY)));
```
@etienneptl added this assert in WT-6682 about 8 months ago.  Etienne, would this be a safe change?

etienneptl replied:
Yes, @ddanderson, I believe this is safe - however, can we ensure the metadata cursor is read-only?

ddanderson replied:
@jeremythorp, assuming there's a way to fix the assert check in `__wt_open_cursor` so it wouldn't fire, is the code improved by removing `session->hs_cursor_counter != 0` check?  Do we check more cases than we would have?

If so, then we could pursue fixing the assert.   @etienneptl, there could be a way to ensure this is a readonly cursor if that is important for correctness. I.e. (would something bad happen if it were not a readonly metadata cursor).  It's a little awkward to check for readonly-ness because at the point of the WT_ASSERT in `__wt_open_cursor`, there are ways to do it, but add a little complexity.  Let's call this "complex assert".

So I think there are three ways out 

1.  keep the code in the PR (maybe it doesn't check some cases we should??), leave cursor_open alone
2. tighten up the ASSERT in meta_ckpt and loosen the ASSERT in cursor open as suggested in code in my comment (maybe it doesn't protect against writeable metadata cursor opens, is that important??)
3. tighten up the ASSERT in meta_ckpt, and use the "complex assert" (I'll describe in another comment).

ddanderson replied:
The "complex assert" would move first move the assert to the end of `__wt_open_cursor`.  So there would be an `err:` label and all the returns (including return(0)) would jump to it.
```
err:
#ifdef WT_DIAGNOSTIC
     if (ret == 0)
         readonly = WT_CURSOR_IS_READONLY(*cursorp);
     else
#define WT_CURSOR_IS_READONLY(cursor)        ((cursor)->insert == __wt_cursor_notsup)

jeremythorp replied:
Hi @ddanderson , I don't like the idea of the added complexity either, so I'd prefer an alternative. 
@etienneptl , do you have any thoughts on this? Thanks, Jeremt

etienneptl replied:
@ddanderson, you said:

> That is, _wt_meta_ckptlist_set on file X can apparently be called while a history store cursor is open

Do we know why we have a HS cursor open in this code flow now? I am not understanding this bit.

ddanderson replied:
@etienneptl, 
> Do we know why we have a HS cursor open in this code flow now? I am not understanding this bit.

I'd guess that this is a checkpoint thread, and perhaps is in the middle of reconciliation, which will need to update the history store.

etienneptl replied:
The questions I have now:

- If we go with @jeremythorp's suggestion, we skip the HS table given the comment line 1482. Does it mean we skip that table while we should process it?
- If we update the other assertion, does it means we process the HS and that's what we want?

etienneptl replied:
After re-reading the code, here is my understanding:

- If we skip the HS, we are not validating the blocks associated with it which imo is wrong. If they are invalid and don't realise it, if the backup is used, it will fail at restoring the right version of the HS. I am against skipping the HS file.
- The new code from the PR does the following as part of checkpoint (below is a stack trace from a failure):

```
__txn_checkpoint
__checkpoint_apply_to_dhandles
__checkpoint_tree_helper
__checkpoint_tree
__wt_meta_ckptlist_set ---> Updated function where we have an issue with an assertion down the road
__check_backup_blocks
__get_blkmods
__wt_metadata_cursor_open
__wt_open_cursor
```

The checkpoint does open a hs cursor before getting to `__wt_meta_ckptlist_set`. We can see that at the end, we open a metadata cursor file. As pointed out by @ddanderson [here](https://github.com/wiredtiger/wiredtiger/pull/9437/files#r1410814813), the assertion should be updated so it allows that code flow when we are opening a new cursor on the metadata file despite having a HS cursor already. I wonder if we should allow this only when a checkpoint is running?

jeremythorp replied:
Hi @ddanderson , yes this is the main issue holding up the PR, so that's why I'm keen to resolve it. @etienneptl and and I are investigating what it would take to make the code work if we remove the 'skip history store' step from the checks - we'll keep you posted.

jeremythorp replied:
Hi @etienneptl and @ddanderson , I've run a series of patch builds with the History Store check removed. I saw a number of failures in the assert in session_api.c. The details are in this [JIRA comment](https://jira.mongodb.org/browse/WT-11178?focusedCommentId=6180603&focusedId=6180603&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-6180603).

jeremythorp replied:
@etienneptl , thanks very much for your help. The next step is to adjust the ASSERT to be ok if we have both a metadata cursor AND a checkpoint is running. I'll make the change tomorrow, test it, and let you know what happens.

etienneptl replied:
@jeremythorp and I had a discussion and we agreed on updating the assertion with the following:
```
diff --git a/src/session/session_api.c b/src/session/session_api.c
index 246c7e879..d4c4f08e0 100644
--- a/src/session/session_api.c
+++ b/src/session/session_api.c
@@ -693,10 +693,13 @@ __wt_open_cursor(WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *owner, co
      * There are two exceptions to this rule:
      *  - Verifying the metadata through an internal session.
      *  - The btree is being verified.
+     *  - Opening the metadata file itself while performing a checkpoint.
      */
     WT_ASSERT(session,
-      strcmp(uri, WT_HS_URI) == 0 || session->hs_cursor_counter == 0 ||
-        F_ISSET(session, WT_SESSION_INTERNAL) ||
+      strcmp(uri, WT_HS_URI) == 0 ||
+        (strcmp(uri, WT_METADATA_URI) == 0 &&
+          __wt_atomic_loadvbool(&txn_global->checkpoint_running)) ||
+        session->hs_cursor_counter == 0 || F_ISSET(session, WT_SESSION_INTERNAL) ||
         (S2BT_SAFE(session) != NULL && F_ISSET(S2BT(session), WT_BTREE_VERIFY)));
 
     /* We do not cache any subordinate tables/files cursors. */
```

ddanderson replied:
Sounds good.

etienneptl replied:
Looks like the check is wrong, it should be `WT_METAFILE_URI` instead.

jeremythorp replied:
Now fixed!

sueloverso commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,137 @@ __wt_meta_ckptlist_update_config(
+    WT_CLEAR(*output_bitmap);
+
+    for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk) {
+        if (!F_ISSET(blk, WT_BLOCK_MODS_VALID))
```
This clause can be removed and the lop further simplified by just:
```
if (F_ISSET(blk, WT_BLOCK_MODS_VALID) && (strcmp(...)))
    WT_RET(...)
```

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,151 @@ __wt_meta_ckptlist_update_config(
 }
 
+/*
+ * ckpt_blkmod_to_item --
```
Done!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,151 @@ __wt_meta_ckptlist_update_config(
+     */
+    if (!valid) {
+        return (0);
+    }
```
Done!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1270,31 +1416,69 @@ int
 err:
     __wt_scr_free(session, &buf);
+    __wt_buf_free(session, &file_blkmods_buffer);
+    __wt_buf_free(session, &checkpoint_blkmods_buffer);
```
Done!

ddanderson commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,133 @@ __wt_meta_ckptlist_update_config(
+        ret = 0;
+
+err:
+    WT_RET(metadata_cursor->close(metadata_cursor));
```
I'd use `WT_TRET` rather than `WT_RET` since we're after the `err:` label.  It doesn't _appear_ to matter in this particular case, but it's a good coding hygene, since with a TRET, any future cleanup code that is added after this line will still get done in error paths.  (I.e. WT can deal with errors and not have resource leaks).

jeremythorp replied:
That's a good suggestion - I've make the change.

ddanderson commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,133 @@ __wt_meta_ckptlist_update_config(
+        /* Loop through the incremental backup blocks data looking for the correct id */
+        while ((ret = __wt_config_next(&blkconf, &blocks_key, &blocks_value)) == 0) {
+            if (blocks_value.len > 0) {
+                /* Have we found the blocks data for the correct key (which has id 'id')? */
```
The code changes you made, make this code much easier to understand.  Given that, I don't think this comment is necessary, nor the "Keep searching" comment, they're pretty much covered by the comment at line 1319. `/* Loop through... */`

jeremythorp replied:
Now removed.

ddanderson reviewed (commented):
@jeremythorp, Here are a few comments on your latest changes.  Other than that, we still have the open issue about the `session->hs_cursor_counter != 0)` check.  That needs a decision - I think that's mostly between you and @etienneptl, I don't know deeply about what's being lost or gained either way.  Beyond that, the PR looks good to me, I'll still hold my approval until the open issue is resolved.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,133 @@ __wt_meta_ckptlist_update_config(
+                if ((ret == 0) && (blocks.len > 0)) {
+                    ret = __wt_nhex_to_raw(session, blocks.str, blocks.len, output_item);
+                    break;
+                }
```
That's a good suggestion - I've make the change.

jeremythorp commented:
@ddanderson and @etienneptl , thanks for your comments and suggestions for handling that assert. Like @ddanderson , I'm not keen on the 'complex assert' ideas as the complexity looks less than ideal. I'm still working on a solution - I'll get back to you with further thoughts/questions.

sueloverso reviewed (commented):
There are still outstanding comments from my earlier review.

jeremythorp commented:
Hi @sueloverso , thank you again for your comments and I'm still working through the remaining ones.

marcbutler reviewed (changes requested):
Several of the questions/conversations I have raised are still unanswered.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,152 @@ __wt_meta_ckptlist_update_config(
+
+/*
+ * check_incorrect_modified_bits --
+ *     This function takes as input two bitmaps (in WT_ITEMs), an original and a new. If any bits in
```
I've made the suggested change, noting that in this context 'bit' and 'bits' are both grammatically correct in this sentence.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,152 @@ __wt_meta_ckptlist_update_config(
+    *ok = false;
+
+    if ((original_bitmap == NULL) || (new_bitmap == NULL))
+        WT_RET(EINVAL);
```
That makes sense - I've made the change.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,152 @@ __wt_meta_ckptlist_update_config(
+get_blkmods(WT_SESSION_IMPL *session, const char *uri, const char *id, WT_ITEM *output_item)
+{
+    WT_CONFIG blkconf;
+    WT_CONFIG_ITEM blocks, key, backup_config_value, blocks_key, blocks_value;
```
Done.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,152 @@ __wt_meta_ckptlist_update_config(
+    WT_CLEAR(backup_config_value);
+    WT_CLEAR(blocks_key);
+    WT_CLEAR(blocks_value);
+    WT_CLEAR(*output_item);
```
Done.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1273,28 +1462,45 @@ __wt_meta_ckptlist_set(
+            if (strcmp(filename, WT_HS_URI) == 0 || session->hs_cursor_counter != 0) {
+                /* Skip history store tables */
+            } else if (F_ISSET(dhandle, WT_DHANDLE_IS_METADATA)) {
+                /* Skip */
```
```suggestion
                /* Skip as there are no backup blocks allocated for the metadata file */
```

jeremythorp commented:
Hi @ddanderson , @etienneptl , @marcbutler and @sueloverso , I've made a number of changes across the recent commits, including

- Revamped and simplified the unit testing - @marcbutler thanks for the feedback and hopefully this now makes more sense.
- Removed the skipping of history store tables, and fixed up the asserts to make this possible. @etienneptl and @ddanderson , thanks for your help on this.
- Various code tidy-ups and improvements, many of which were suggested by @sueloverso - thank you for that. 

One final, long-running test, which has been a good detector of the history store assert issues is still running as I write this. Assuming that test passes, then I think the PR is ready for re-review.

Thanks, Jeremy

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
 
+/*
+ * __ckpt_extract_blkmod_bitmap --
+ *     Extracts the blkmod info in a checkpoint into a bitmap in a WT_ITEM.
```
The sentence looks odd to me: in... into... in.
```suggestion
 *     Extracts the blkmod info in a checkpoint into a bitmap.
```
If we really want to mention `WT_ITEM`, I'd suggest:
```suggestion
 *     Extracts the blkmod info in a checkpoint into a WT_ITEM.
```

jeremythorp replied:
I've changed it to "Extracts the blkmod info in a checkpoint into a bitmap within a WT_ITEM." How does that sound?

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+
+    WT_CLEAR(*output_bitmap);
+
+    for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk)
```
Should `&ckpt->backup_blocks[0]` be `&ckpt->backup_blocks[i]`?

jeremythorp replied:
Hi @etienneptl . `&ckpt->backup_blocks[0]` is correct in this case as it initialising `blk` at the start of the loop, and then `blk` is incremented each time around. `i` is used simply to count the number of loop iterations rather than as an index.

etienneptl replied:
Thank you, I missed the {{++blk}} bit.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+
+/*
+ * __get_blkmods --
+ *     Extracts the blkmod info from the metadata about a file+id into a bitmap in a WT_ITEM.
```
Same comment here regarding the description. You can leave it as is if you decide to ignore my first comment.

jeremythorp replied:
See above.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+    WT_ERR(
+      __wt_config_getones(session, file_config, "checkpoint_backup_info", &backup_config_value));
+
+    if (backup_config_value.len > 0) {
```
Should we early exit here by jumping to the err label instead of having a double indent?
```
if (backup_config_value.len == 0 || backup_config_value.type != WT_CONFIG_ITEM_STRUCT)
     goto done;
```
Note that `done` is not existing, we could do `goto err;` but it's not an error... So it could be misleading.

jeremythorp replied:
Hi @etienneptl , I prefer to avoid `goto` statements whenever possible, so I prefer the `if` statement as it is rather than adding another `goto`.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+
+            /* Loop through the incremental backup blocks data looking for the correct id */
+            while ((ret = __wt_config_next(&blkconf, &blocks_key, &blocks_value)) == 0) {
+                if (blocks_value.len > 0) {
```
```suggestion
                if (blocks_value.len == 0)
                    continue;
```

jeremythorp replied:
Hi @etienneptl , I think that change (including reducing the indent of the next lines) makes the code easier to read - I'll make the change.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+                                break;
+                            }
+                        }
+                        WT_ERR_NOTFOUND_OK(ret, false);
```
You can move this just after the call to `__wt_config_subgets` line 1325 and remove the check `if (ret == 0) {.

jeremythorp replied:
Hi @etienneptl , because `__wt_config_subgets` could return `WT_NOTFOUND`, we have to confirm that `ret == 0` before trying to extract the block data. If the `WT_ERR_NOTFOUND_OK()` is moved earlier then we won't know if the blocks were found in the config, so I believe it's in the right place.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+                        ret = __wt_config_subgets(session, &blocks_value, "blocks", &blocks);
+                        if (ret == 0) {
+                            if (blocks.len > 0) {
+                                ret =
```
What are we doing with this ret? Is it needed? Looks like we need `WT_ERR` instead.

jeremythorp replied:
Hi @etienneptl , that makes sense - I've made the change.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+    WT_NOT_READ(index, 0);
+    original_ptr = NULL;
+    new_ptr = NULL;
+    WT_NOT_READ(partial_result, 0);
```
Is `WT_NOT_READ` needed?

jeremythorp replied:
We need it because otherwise we get the following errors from the clang analyser:

```
[2024/03/17 19:54:29.018] /data/mci/8e9fbae01b34d2314f0ee58c6aa5e59e/wiredtiger/src/meta/meta_ckpt.c:1362:5: warning: Value stored to 'index' is never read [deadcode.DeadStores]
[2024/03/17 19:54:29.018]     index = 0;
[2024/03/17 19:54:29.018]     ^       ~
[2024/03/17 19:54:29.018] /data/mci/8e9fbae01b34d2314f0ee58c6aa5e59e/wiredtiger/src/meta/meta_ckpt.c:1365:5: warning: Value stored to 'partial_result' is never read [deadcode.DeadStores]
[2024/03/17 19:54:29.018]     partial_result = 0;
```

etienneptl replied:
I don't understand though, it is read [here](https://github.com/wiredtiger/wiredtiger/pull/9437/files#diff-3be0ed5707ef4b0f982eac862ff81534805bd6a7c363fa4241d8a09534e1cb20R1397).

jeremythorp replied:
I also don't understand why the errors are generated in this case because, as you say, the variables are read. My current theory is that the compiler thinks there are some paths where the written values are never subsequently read, but that requires further investigation.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+                  &file_blkmods_buffer, checkpoint_blkmods_buffer, &blkmods_are_ok);
+
+                if ((ret != 0) || !blkmods_are_ok) {
+                    WT_ASSERT(session, false);
```
Why not a `WT_ASSERT_ALWAYS` instead?

jeremythorp replied:
We need it because otherwise we get the following errors from the clang analyser:

```
[2024/03/17 19:54:29.018] /data/mci/8e9fbae01b34d2314f0ee58c6aa5e59e/wiredtiger/src/meta/meta_ckpt.c:1362:5: warning: Value stored to 'index' is never read [deadcode.DeadStores]
[2024/03/17 19:54:29.018]     index = 0;
[2024/03/17 19:54:29.018]     ^       ~
[2024/03/17 19:54:29.018] /data/mci/8e9fbae01b34d2314f0ee58c6aa5e59e/wiredtiger/src/meta/meta_ckpt.c:1365:5: warning: Value stored to 'partial_result' is never read [deadcode.DeadStores]
[2024/03/17 19:54:29.018]     partial_result = 0;
```

jeremythorp replied:
I'm happy change it. Because of the subsequent `WT_ERR_PANIC()`, I think either `WT_ASSERT` or `WT_ASSERT_ALWAYS` is ok here.

etienneptl replied:
Should we keep both though?

jeremythorp replied:
I think the panic is the right option and I'm ok with having either the WT_ASSERT or the WT_ASSERT_ALWAYS with it. I think there was a comment suggesting the WT_ASSERT be added much earlier in the discussion, but I don't seem to be able to find it now.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+
+                if ((ret != 0) || !blkmods_are_ok) {
+                    WT_ASSERT(session, false);
+                    WT_ERR_PANIC(session, WT_PANIC,
```
I think it would be worth it to have information about the checkpoint that is being processed in the message for easier diagnosis.

jeremythorp replied:
That's a good idea - I'll make the change.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1273,28 +1462,43 @@ __wt_meta_ckptlist_set(
+
             WT_ERR(__wt_ckpt_blkmod_to_meta(session, buf, ckpt));
 
+            if (F_ISSET(dhandle, WT_DHANDLE_IS_METADATA)) {
+                /* Skip as there are no backup blocks allocated for the metadata file */
+            } else {
+                WT_ERR(__check_backup_blocks(session, ckpt, &checkpoint_blkmods_buffer, filename));
+            }
```
```suggestion
            /* Skip the metadata file as there are no backup blocks allocated for it. */
            if (!F_ISSET(dhandle, WT_DHANDLE_IS_METADATA))
                WT_ERR(__check_backup_blocks(session, ckpt, &checkpoint_blkmods_buffer, filename));
```

jeremythorp replied:
Hi @etienneptl , while my version is longer, I think it's easier to read.

etienneptl reviewed (commented):
Thank you for all the changes and the great comments - makes it easy to follow. I have a few observations/comments I am happy to discuss.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+    metadata_cursor->set_key(metadata_cursor, uri);
+    WT_ERR(metadata_cursor->search(metadata_cursor));
+    WT_ERR(metadata_cursor->get_value(metadata_cursor, &file_config));
+    WT_ERR(
```
Can we get `WT_NOTFOUND` here? Is it possible or would that be an issue?
If it's possible, the `WT_ERR` would be incorrect I believe.

jeremythorp replied:
Hi @etienneptl , that's a great question, and I think it's at least theoretically possible. The solution is to move the 
`    if (ret == WT_NOTFOUND)
        ret = 0;`
after the 'err' label.

etienneptl replied:
I am afraid this could catch unwanted errors if this function ever changes.
I would prefer:
```
if (ret == WT_NOTFOUND) {
  /* Nothing to check. we are done. */
  ret = 0;
  goto err;
}
```

etienneptl replied:
Since 3952c6e3cbccedccc8f6ba3f3303528ebef1d4fa, you are now hiding a potential error from the call `WT_ERR(metadata_cursor->search(metadata_cursor));`.

etienneptl replied:
@sueloverso, what are your thoughts on this?

jeremythorp replied:
@etienneptl and @sueloverso , thanks for the feedback. I've moved the check for `WT_NOTFOUND` back to before the `err` label, and added an assert for WT_NOTFOUND on the result of `metadata_cursor->search` as suggested.

jeremythorp commented:
@etienneptl , thanks for your feedback, and I've made a number of improvements in response.
@sueloverso , @ddanderson and @marcbutler , when you get the chance to re-review the PR, I'd very much appreciate your comments.

sueloverso commented on src/meta/meta_ckpt.c:
```diff
@@ -1273,28 +1465,43 @@ __wt_meta_ckptlist_set(
+            if (F_ISSET(dhandle, WT_DHANDLE_IS_METADATA)) {
+                /* Skip as there are no backup blocks allocated for the metadata file */
+            } else {
+                WT_ERR(__check_backup_blocks(session, ckpt, &checkpoint_blkmods_buffer, filename));
```
Should all of this be under DIAGNOSTIC? I'm not sure we want to be performing this verification on production builds on every checkpoint? On the other hand, this is not a performant path. I don't know the answer but need to ask the question!

jeremythorp replied:
I think it's important to perform these checks all the time (unless there's a perf impact) so that we are verifying on production environments, just in case a problem slips in down the track.

sueloverso commented on src/session/session_api.c:
```diff
@@ -684,19 +684,24 @@ __wt_open_cursor(WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *owner, co
     /*
      * We should not open other cursors when there are open history store cursors in the session.
-     * There are two exceptions to this rule:
+     * There are three exceptions to this rule:
```
Maybe future-proof by changing `two` to `a few` or `some`

jeremythorp replied:
Done!

sueloverso commented on test/csuite/incr_backup/main.c:
```diff
@@ -523,63 +523,33 @@ check_backup(uint32_t backup_iter, TABLE_INFO *tinfo)
+    int ncheckpoints, nreopens;
+    WT_FILE_COPY_OPTS copy_opts;
     char backup_src[1024], conf[1024], home[1024];
-    bool preserve;
```
Style: Can you undo the reordering here? The original was per style guidelines and this now is not. Was there a reason to rearrange them?

jeremythorp replied:
My apologies - I'm not sure what happened there. Now fixed.

jeremythorp replied:
Actually, something odd is going on with s_all. It is insisting on the following change:
```
int
 main(int argc, char *argv[])
 {
-    int ch;
-    bool preserve;
     WT_RAND_STATE rnd;
     uint64_t seed_param;
+    int ch;
     const char *working_dir;
+    bool preserve;
 
     preserve = false;
     seed_param = 0;`
```

jeremythorp replied:
The order of the locals in both main() and run_test() should now be as expected by s_all.

jeremythorp replied:
I'm unsure either, as s_all was happy about the incorrect location of WT_FILE_COPY_OPTS. I've moved it to the correct spot, and s_all is still happy. Perhaps there are some limitations in the s_all checks themselves.

sueloverso commented on test/evergreen/code_coverage/code_coverage_config.json:
```diff
@@ -145,6 +145,8 @@
     "test/csuite/wt3120_filesys/test_wt3120_filesys -b $(pwd)",
     "test/csuite/wt8659_reconstruct_database_from_logs/test_wt8659_reconstruct_database_from_logs",
+    "test/csuite/incr_backup/test_incr_backup -S 123456789",
+    "test/csuite/incr_backup/test_incr_backup -S 0x9b1bde3f111fe316",
```
It feels like maybe these cases should be also added to `test/csuite/incr_backup/smoke.sh`?

jeremythorp replied:
Hi @sueloverso , that's a great suggestion, but unless I've missed something is actually not necessary. 
Here's why: With my changes and when not given a seed, incr_backup itself now runs the test three times: once each with those specific seeds, and once with a random seed. That means it will always trigger a failure without your fix to WT-10551 (because one of those seeds generates a failure). The reason I'm running with a seed (twice) in the code coverage is to specifically avoid the non-deterministic random seed test.
This means that the default (as executed by smoke.sh) will run with those two seeds and a random one already.

etienneptl commented on src/meta/meta_ckpt.c:
```diff
@@ -1273,28 +1465,43 @@ __wt_meta_ckptlist_set(
             WT_ERR(__wt_ckpt_blkmod_to_meta(session, buf, ckpt));
 
+            if (F_ISSET(dhandle, WT_DHANDLE_IS_METADATA)) {
+                /* Skip as there are no backup blocks allocated for the metadata file */
```
Suggested [here](https://github.com/wiredtiger/wiredtiger/pull/9437#discussion_r1525918351) too

jeremythorp replied:
Updated!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+ */
+static int
+__ckpt_extract_blkmod_bitmap(
+  WT_SESSION_IMPL *session, WT_CKPT *ckpt, const char *id, WT_ITEM *output_bitmap)
```
That makes sense - I've renamed it.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+ *     Extracts the blkmod info from the metadata about a file+id into a bitmap within a WT_ITEM.
+ */
+static WT_INLINE int
+__get_blkmods(WT_SESSION_IMPL *session, const char *uri, const char *id, WT_ITEM *output_item)
```
Makes sense - I've renamed both.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+}
+
+/*
+ * __check_incorrect_modified_bits --
```
Done!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+    WT_ERR(
+      __wt_config_getones(session, file_config, "checkpoint_backup_info", &backup_config_value));
+
+    if (backup_config_value.len > 0) {
```
Hi @sueloverso , I believe we do need those conditionals.
I think the equivalent assert to be able to remove them would be `WT_ASSERT(session, (backup_config_value.len > 0) && (backup_config_value.type == WT_CONFIG_ITEM_STRUCT));`
I tried running the tests with that asset, and I quickly get an assert failure.
What do you think?

jeremythorp replied:
Thanks @sueloverso , I've refactored the if statement as suggested.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+
+    WT_NOT_READ(index, 0);
+    original_ptr = NULL;
+    new_ptr = NULL;
```
On this one, my preference is to leave the two lines seperate, as I think it is easier to read.

sueloverso reviewed (approved):
I have a few more minor comments but no need to review again.

marcbutler commented on test/unittest/tests/wrappers/connection_wrapper.h:
```diff
@@ -37,11 +37,18 @@ class ConnectionWrapper {
     WT_CONNECTION *getWtConnection() const;
 
+    void
+    clearDoCleanup()
```
I find the function name and the state mutating semantics of this difficult to follow.
It is probably beyond the scope of this PR but I think setting up the db home dir should be separate from this class: possibly an object that is passed into this class: that way it could take a parameter in its ctor that indicates whether or not the dir should be cleaned up.

jeremythorp replied:
Hi @marcbutler , I think that further work can be done to improve the wrapper classes, and I think that is outside the scope of this ticket, but well worth doing.

marcbutler replied:
agreed

marcbutler commented on test/unittest/tests/backup/test_incremental_blkmods.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+static uint8_t
+get_hex_value_from_string(std::string const &source_string, size_t index)
```
Have you considered a single function for transforming the string into a byte vector?
As I see it one of the downsides of this function is that is that if there is a non-hex digit(s) in the string it will not be discovered (if at all) until queried. Compiling the entire vector up front provides a single point of failure.

Additionally if an invalid index is passed to this function it returns zero rather than signalling an error via an exception -- is this intended? If so why is an invalid index not considered an error?

jeremythorp replied:
The case where the function returned 0 was correct, as that's because no data can be interpreted as 0. However, I agree that is was unclear. 

I've following your suggestion of revamping the code to convert the hex strings up front and to make the invalid character and different string length cases clearer in the code and comments.

marcbutler commented on test/unittest/tests/wrappers/item_wrapper.cpp:
```diff
@@ -9,7 +9,7 @@
-item_wrapper::item_wrapper(const char *string) : _string(string)
+item_wrapper::item_wrapper(std::string const &string) : _string(string)
 {
     _item.data = _string.c_str();
```
My reading is that the `WT_ITEM` this class wraps is at least conceptually mutable.
However modifying the memory returned by `c_str()` is undefined behavior *[string.accessors]* in the C++17 standard.

jeremythorp replied:
Hi @marcbutler , you are correct, and I agree that is not currently made clear.
The `item_wrapper` as currently implemented is intended to wrap only constant, read-only strings. I've added a comment to `item_wrapper.h` explaining that, and we should consider enhancing `item_wrapper` in future tickets.

marcbutler replied:
agreed

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,198 @@ __wt_meta_ckptlist_update_config(
+                if (WT_STRING_MATCH(id, blocks_key.str, blocks_key.len)) {
+                    /* We've found the right blocks so read the bit pattern into output_item */
+                    ret = __wt_config_subgets(session, &blocks_value, "blocks", &blocks);
+                    if (ret == 0) {
```
Done!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1273,28 +1471,35 @@ __wt_meta_ckptlist_set(
 
 err:
     __wt_scr_free(session, &buf);
+
```
Done!

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1273,28 +1471,35 @@ __wt_meta_ckptlist_set(
     WT_DECL_ITEM(buf);
     WT_DECL_RET;
-    const char *fname;
+    const char *filename;
```
I've changed it back.

jeremythorp commented on src/meta/meta_ckpt.c:
```diff
@@ -1262,6 +1262,204 @@ __wt_meta_ckptlist_update_config(
+            break;
+        }
+
+    return (0);
```
That's a great question - the calling code (in `__ckpt_check_backup_blocks`) checks the content of the returned `output_bitmap`. If it's empty, then nothing was found and the caller keeps going.

etienneptl reviewed (approved):
LGTM!

jeremythorp commented:
@sueloverso and @etienneptl , thanks very much for your LGTMs!
@marcbutler , thanks for the feedback on the unit tests - I've revamped the code in line with your suggestions and added more comments to make it clearer. Please let me know what you think.

marcbutler reviewed (approved):
lgtm
